src/Storages/StorageMaterializedView.cpp

```diff
@@ -247,7 +261,54 @@ bool StorageMaterializedView::optimize(
     checkStatementCanBeForwarded();
     auto storage_ptr = getTargetTable();
     auto metadata_snapshot = storage_ptr->getInMemoryMetadataPtr();
-    return getTargetTable()->optimize(query, metadata_snapshot, partition, final, deduplicate, deduplicate_by_columns, cleanup, local_context);
+    return storage_ptr->optimize(query, metadata_snapshot, partition, final, deduplicate, deduplicate_by_columns, cleanup, local_context);
+}
+
+StorageID StorageMaterializedView::createFreshTable() const
+{
+    auto inner_table_id = getTargetTableId();
+    auto new_table_name = ".tmp" + generateInnerTableName(getStorageID());
+
+    auto db = DatabaseCatalog::instance().getDatabase(inner_table_id.database_name);
+
+    auto create_table_query = db->getCreateTableQuery(inner_table_id.table_name, getContext());
+    auto & create_query = create_table_query->as<ASTCreateQuery &>();
+    create_query.setTable(new_table_name);
+    create_query.setDatabase(db->getDatabaseName());
+    create_query.create_or_replace = true;
+    create_query.replace_table = true;
+    create_query.uuid = UUIDHelpers::Nil;
+
+    auto create_ctx = Context::createCopy(getContext());
+    InterpreterCreateQuery create_interpreter(create_table_query, create_ctx);
+    create_interpreter.setInternal(true);
+    create_interpreter.execute();
+
+    return DatabaseCatalog::instance().getTable({create_query.getDatabase(), create_query.getTable()}, getContext())->getStorageID();
```

Comment:
nit: What if the user manually drops the table (is it possible?) between these two lines? Worth at least checking for nullptr here, to not crash?